crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2589,6 +2600,21 @@ impl<'db> TypeInferenceBuilder<'db> {
                 }
             }
 
+            (Type::BooleanLiteral(_), Type::BooleanLiteral(_), op) => match op {
+                ruff_python_ast::Operator::Add
+                | ruff_python_ast::Operator::Sub
+                | ruff_python_ast::Operator::Mult
+                | ruff_python_ast::Operator::Mod
+                | ruff_python_ast::Operator::Pow
+                | ruff_python_ast::Operator::LShift
+                | ruff_python_ast::Operator::RShift
+                | ruff_python_ast::Operator::BitXor
+                | ruff_python_ast::Operator::BitAnd
+                | ruff_python_ast::Operator::FloorDiv
+                | ruff_python_ast::Operator::MatMult => KnownClass::Int.to_instance(self.db),
+                ruff_python_ast::Operator::BitOr => KnownClass::Bool.to_instance(self.db),
+                ruff_python_ast::Operator::Div => KnownClass::Float.to_instance(self.db),
+            },
```

Comment:
`BitOr` of two `BooleanLiteral` needs special handling here; I'd give it its own case, and aim to return the correct `BooleanLiteral` for all four cases (should be easy to do as one line by just ORing the internal rust bools from each operand).

For everything else, I'd just have patterns for `(left, BooleanLiteral(b), op)` and for `(BooleanLiteral(b), right, op)` where all those cases do is construct the correct 1-or-0 `IntLiteral` for the `BooleanLiteral` (see above where that's already done in `infer_unary_expression`) and then re-run this match using that int literal in place of the boolean literal.

This will require extracting an `infer_binary_expression_type` function, which takes `left_ty: Type, right_ty: Type, op` arguments and just contains the match statement, so that you can recursively call it with the new types. You can see where we already did that in https://github.com/astral-sh/ruff/pull/13800, but it's fine to do it here too, one can easily rebase on the other.

Reply:
Thanks this definitely works better than what I did.

Just one note, I was thinking to also move the zero division check to the recursive call so if a non trivial zero division is covered in the recursions it's also reported but I found that cases like `True / False` will become `Cannot divide object of type `Literal[1]` by zero` so I left it in the parent function.
It makes sense because we can know anything that will become zero division error.